File: app/src/main/java/com/github/marlonlom/demos/barcodescans/CameraBarcodeScannerFragment.java
package com.github.marlonlom.demos.barcodescans;

import android.Manifest;
import android.app.Activity;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Bundle;
import android.os.Handler;
import android.provider.MediaStore;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.Snackbar;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.ContextCompat;
import android.util.SparseArray;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;

import com.google.android.gms.vision.Frame;
import com.google.android.gms.vision.barcode.Barcode;

/**
 * Camera Barcode scanner fragment class.
 *
 * @author marlonlom
 */
public final class CameraBarcodeScannerFragment extends BarcodeDetectorFragment {
    public static final int RESULT_INTENT_PHOTO_CAPTURE = 100;
    private static final int PERMISSION_REQUEST_PHOTO_CAPTURE = 200;
    private Button mScanButton;
    private ImageView mBarcodeImage;
    private TextView mResultsText;
    private Uri mImageUri;

    public CameraBarcodeScannerFragment() {
        super();
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
            @Nullable Bundle savedInstanceState) {

        final View rootView = inflater.inflate(R.layout.fragment_tab_camera_scan, container, false);

        mScanButton = (Button) rootView.findViewById(R.id.button_camera_scan);
        mBarcodeImage = (ImageView) rootView.findViewById(R.id.imageview_camera_barcode);
        mResultsText = (TextView) rootView.findViewById(R.id.textview_camera_results);

        setupView();

        return rootView;

    }

    @Override
    protected void setupView() {
        mScanButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                checkCameraPermission();
            }
        });

        final Handler handler = new Handler();
        handler.postDelayed(new Runnable() {
            @Override
            public void run() {
                final int testResourceId =
                        !isDetectorOperational() ? R.string.text_error_no_operational
                                : R.string.textView_waiting;
                mResultsText.setText(testResourceId);
            }
        }, 2000);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
            @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        switch (requestCode) {
            case PERMISSION_REQUEST_PHOTO_CAPTURE:
                if (grantResults.length > 0
                        && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    takePicture();
                } else {
                    mScanButton.setEnabled(false);
                }
                break;
        }
    }

    private void takePicture() {
        mResultsText.setText(R.string.text_initializing);
        Intent takePictureIntent = new Intent(MediaStore.ACTION_IMAGE_CAPTURE);
        if (takePictureIntent.resolveActivity(getActivity().getPackageManager()) != null) {
            startActivityForResult(takePictureIntent, RESULT_INTENT_PHOTO_CAPTURE);
        }
    }

    private void checkCameraPermission() {
        if (ContextCompat.checkSelfPermission(getActivity(), Manifest.permission.CAMERA)
                != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(getActivity(),
                    new String[]{Manifest.permission.CAMERA}, PERMISSION_REQUEST_PHOTO_CAPTURE);
        } else {
            takePicture();
        }
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        switch (requestCode) {
            case RESULT_INTENT_PHOTO_CAPTURE:
                if (resultCode == Activity.RESULT_OK) {
                    processTakenPicture(data);
                }
                break;
        }
    }

    private void processTakenPicture(Intent data) {
        final Bitmap photoBitmap = (Bitmap) data.getExtras().get("data");
        if (photoBitmap != null) {
            mResultsText.setText(R.string.text_camera_scanning);
            mBarcodeImage.setImageBitmap(photoBitmap);
            new Handler().postDelayed(new Runnable() {
                @Override
                public void run() {
                    performBarcodeScanning(photoBitmap);
                }
            }, 2000);

        } else {
            Snackbar.make(mBarcodeImage, "Could not get the picture to scan",
                    Snackbar.LENGTH_SHORT).show();
        }
    }

    private void performBarcodeScanning(final Bitmap bitmap) {
        Frame frame = new Frame.Builder().setBitmap(bitmap).build();
        SparseArray<Barcode> barCodes = getDetector().detect(frame);
        if (barCodes.size() > 0) {
            Barcode thisCode = barCodes.valueAt(0);
            mResultsText.setText(thisCode.rawValue);
        } else {
            mResultsText.setText(R.string.text_error_scanning);
        }
    }
}

File: app/src/main/kotlin/com/github/marlonlom/demos/barcodescans/CameraBarcodeScannerFragment.kt
package com.github.marlonlom.demos.barcodescans

import android.Manifest
import android.app.Activity
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.net.Uri
import android.os.Bundle
import android.os.Handler
import android.provider.MediaStore
import android.support.v4.app.ActivityCompat
import android.support.v4.content.ContextCompat
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import com.google.android.gms.vision.Frame

/**
 * Camera Barcode scanner fragment class.

 * @author marlonlom
 */
class CameraBarcodeScannerFragment : BarcodeDetectorFragment() {
    private var mScanButton: Button? = null
    private var mBarcodeImage: ImageView? = null
    private var mResultsText: TextView? = null
    private val mImageUri: Uri? = null

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {

        val rootView = inflater!!.inflate(R.layout.fragment_tab_camera_scan, container, false)

        mScanButton = rootView.findViewById(R.id.button_camera_scan) as Button
        mBarcodeImage = rootView.findViewById(R.id.imageview_camera_barcode) as ImageView
        mResultsText = rootView.findViewById(R.id.textview_camera_results) as TextView

        setupView()

        return rootView

    }

    override fun setupView() {
        mScanButton!!.setOnClickListener { checkCameraPermission() }

        val handler = Handler()
        handler.postDelayed({
            val testResourceId = if (!isDetectorOperational)
                R.string.text_error_no_operational
            else
                R.string.textView_waiting
            mResultsText!!.setText(testResourceId)
        }, 2000)
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>,
                                            grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        when (requestCode) {
            PERMISSION_REQUEST_PHOTO_CAPTURE -> if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                takePicture()
            } else {
                mScanButton!!.isEnabled = false
            }
        }
    }

    private fun takePicture() {
        mResultsText!!.setText(R.string.text_camera_scanning)
        val takePictureIntent = Intent(MediaStore.ACTION_IMAGE_CAPTURE)
        if (takePictureIntent.resolveActivity(activity.packageManager) != null) {
            startActivityForResult(takePictureIntent, RESULT_INTENT_PHOTO_CAPTURE)
        }
    }

    private fun checkCameraPermission() {
        if (ContextCompat.checkSelfPermission(activity, Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(activity,
                    arrayOf(Manifest.permission.CAMERA), PERMISSION_REQUEST_PHOTO_CAPTURE)
        } else {
            takePicture()
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent) {
        super.onActivityResult(requestCode, resultCode, data)
        when (requestCode) {
            RESULT_INTENT_PHOTO_CAPTURE -> if (resultCode == Activity.RESULT_OK) {
                processTakenPicture(data)
            }
        }
    }

    private fun processTakenPicture(data: Intent) {
        val photoBitmap = data.extras.get("data") as Bitmap
        mResultsText!!.setText(R.string.text_camera_scanning)
        mBarcodeImage!!.setImageBitmap(photoBitmap)
        Handler().postDelayed({ performBarcodeScanning(photoBitmap) }, 2000)
    }

    private fun performBarcodeScanning(bitmap: Bitmap) {
        val frame = Frame.Builder().setBitmap(bitmap).build()
        val barCodes = detector!!.detect(frame)
        if (barCodes.size() > 0) {
            val thisCode = barCodes.valueAt(0)
            mResultsText!!.text = thisCode.rawValue
        } else {
            mResultsText!!.setText(R.string.text_error_scanning)
        }
    }

    companion object {
        val RESULT_INTENT_PHOTO_CAPTURE = 100
        private val PERMISSION_REQUEST_PHOTO_CAPTURE = 200
    }
}
